### PR TITLE
Add DS3231 Component

### PR DESCRIPTION
**Pull request by nuttytree**
## Description:
Adds a DS3231 Real Time Clock Component that includes:

Core component that handles communication with the device
Options on the core component to setup alarms
Options on the core component to set the mode and frequency of the square wave output
Action automations to set the alarms
Trigger automations for when the alarms are triggered
Action automations to reset the alarms after they have been triggered
Action automations to set the mode and frequency of the square wave output
Time component
Action automations to read/write the time from/to the RTC
Temperature sensor

**Pull request in [esphome](https://github.com/esphome/esphome) with YAML changes (if applicable):** esphome/esphome#2501

## Checklist:

  - [X] Branch: `next` is for changes and new documentation that will go public with the next ESPHome release. Fixes, changes and adjustments for the current release should be created against `current`.
  - [X] Link added in `/index.rst` when creating new documents for new components or cookbook.

### CODE REVIEW

File: components/ds3231.rst [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+DS3231 RTC
+==========
+
+.. seo::
+    :description: Instructions for setting up a DS3231 Real Time Clock in ESPHome
+    :image: ds3231.jpg
+
+The ``DS3231`` (`datasheet <https://datasheets.maximintegrated.com/en/ds/DS3231.pdf>`__) is a real time clock device
+that includes 2 alarms, an output that can be triggered by either alarm or output a continuous square wave at 1hz,
+1024hz, 4096hz, or 8192hz, a 32.768kHz output, and a temperature sensor. It uses :ref:`I²C Bus <i2c>` for communication.
```
Comment (oxan) [inline components/ds3231.rst]:
```suggestion
that includes 2 alarms, an output that can be triggered by either alarm or output a continuous square wave at 1 Hz,
1024 Hz, 4096 Hz, or 8192 Hz, a 32.768 kHz output, and a temperature sensor. It uses :ref:`I²C Bus <i2c>` for communication.
```

File: components/ds3231.rst [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+- **id** (*Optional*, :ref:`config-id`): Manually specify the ID used for code generation.
+- **address** (*Optional*, int): The I²C address of the DS3231. Defaults to ``0x68``.
+- **i2c_id** (*Optional*): The I²C Bus ID
+- **update_interval** (*Optional*, :ref:`config-time`): The interval to check the alarm triggers. Defaults to ``60s``.
+- **square_wave_mode** (*Optional*, enum): The mode of the square wave output. Defaults to ``ALARM_INTERRUPT``.
+
+  - ``ALARM_INTERRUPT`` - The square wave output goes high when an alarm with interrupt_enabled set to true is triggered.
+  - ``SQUARE_WAVE`` - The square wave output produces a continuous square wave at the specified frequency.
+
+- **square_wave_frequncy** (*Optional*, enum): The frequency of the square wave output. Defaults to ``8192HZ``.
```
Comment (oxan) [inline components/ds3231.rst]:
```suggestion
- **square_wave_frequency** (*Optional*, enum): The frequency of the square wave output. Defaults to ``8192HZ``.
```

File: components/ds3231.rst [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+  - ``1024HZ``
+  - ``4096HZ``
+  - ``8192HZ``
+
+- **alarm_1** (*Optional*): The configuration for alarm 1.
+
+  - **mode** (**Required**, enum): The mode of the alarm.
+
+    - ``EVERY_SECOND`` - Alarm is triggered every second.
+    - ``MATCH_SECOND`` - Alarm is triggered everytime the second matches the provided value.
```
Comment (oxan) [inline components/ds3231.rst]: "everytime" should be "every time" throughout this file.

File: components/ds3231.rst [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+``ds3231.set_alarm_1`` Action
+-----------------------------
+
+Sets the configuration for alarm 1.
+
+.. code-block:: yaml
+
+    on_...:
+      then:
+        - fan.ds3231.set_alarm_1:
```
Comment (oxan) [inline components/ds3231.rst]: `fan.`? (also below).

**Review by oxan — CHANGES_REQUESTED** [review]

File: components/ds3231.rst (hunk 1 above)
Comment (nuttytree) [reply components/ds3231.rst]: Updated

File: components/ds3231.rst (hunk 2 above)
Comment (nuttytree) [reply components/ds3231.rst]: Fixed

File: components/ds3231.rst (hunk 4 above)
Comment (nuttytree) [reply components/ds3231.rst]: Fixed

File: components/ds3231.rst (hunk 3 above)
Comment (nuttytree) [reply components/ds3231.rst]: Fixed

**Review by oxan — APPROVED** [review]
Ready to be merged when parent PR is.

**Comment by nuttytree** [comment]
Sorry for the delay on getting this tested, I was out with Covid for 3 weeks.  Trying to get caught back up so I can get this tested.

**Pull request closed without merging**